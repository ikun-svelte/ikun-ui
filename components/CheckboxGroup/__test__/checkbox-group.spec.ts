import { tick } from 'svelte';
import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import KCheckboxGroup from '../src';
import KGroupBase from './fixture/group.base.svelte';
import KGroupChange from './fixture/group.change.svelte';
import KGroupTrigger from './fixture/group.trigger.svelte';
import KGroupBaseDisabled from './fixture/group.base.disabled.svelte';
import KGroupChangeDisabled from './fixture/group.change.disabled.svelte';
import KGroupTriggerDisabled from './fixture/group.trigger.disabled.svelte';
import KGroupModelDisabled from './fixture/group.model.disabled.svelte';
import KGroupSize from './fixture/group.size.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KCheckboxGroup', () => {
	test('props: cls', async () => {
		const instance = new KCheckboxGroup({
			target: host,
			props: {
				cls: 'k-checkbox-group--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-checkbox-group--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size', async () => {
		const instance = new KGroupSize({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const smEls = host.querySelector('#group-sm');
		expect(smEls!.querySelectorAll('.k-checkbox--sm').length).toBe(3);
		const mdEls = host.querySelector('#group-md');
		expect(mdEls!.querySelectorAll('.k-checkbox--md').length).toBe(5);
		const lgEls = host.querySelector('#group-lg');
		expect(lgEls!.querySelectorAll('.k-checkbox--lg').length).toBe(2);
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is not disabled, the value of the checkbox can be initialized.', async () => {
		const instance = new KGroupBase({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is not disabled, the value changes and the status of the checkbox changes.', async () => {
		const instance = new KGroupChange({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const btn = host.querySelector('#group_change_btn') as HTMLElement;
		btn!.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is not disabled, when a value of the checkbox changes, the group value is triggered.', async () => {
		const instance = new KGroupTrigger({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const divEl = host.querySelector('#group_trigger_value');
		expect(divEl?.innerHTML).toBe('3');
		inputEls[0].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3,1');
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		inputEls[1].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3,1,2');
		inputEls[0].click();
		await tick();
		inputEls[1].click();
		await tick();
		inputEls[2].click();
		await tick();
		expect(divEl?.innerHTML).toBe('');
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is disabled, the value of the checkbox can be initialized.', async () => {
		const instance = new KGroupBaseDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		expect(host.innerHTML.includes('k-checkbox--box__disabled')).toBeTruthy();
		expect(host.innerHTML.match(/k-checkbox--box__disabled/g)?.length === 3).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is disabled, the value changes and the status of the checkbox not changes.', async () => {
		const instance = new KGroupChangeDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const btn = host.querySelector('#group_change_disabled_btn') as HTMLElement;
		btn!.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is disabled, when a value of the checkbox changes, the group value is not triggered.', async () => {
		const instance = new KGroupTriggerDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const divEl = host.querySelector('#group_trigger_value');
		expect(divEl?.innerHTML).toBe('3');
		inputEls[0].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		inputEls[1].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3');
		inputEls[0].click();
		await tick();
		inputEls[1].click();
		await tick();
		inputEls[2].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3');
		expect(host.innerHTML).matchSnapshot();
	});

	test('When the disabled value changes, the checkbox value can be synchronized.', async () => {
		const instance = new KGroupModelDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const btn = host.querySelector('#group_change_value_btn') as HTMLElement;
		btn!.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();

		const btn2 = host.querySelector('#group_change_disabled_btn') as HTMLElement;
		btn2!.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();

		expect(host.innerHTML).matchSnapshot();
	});
});
