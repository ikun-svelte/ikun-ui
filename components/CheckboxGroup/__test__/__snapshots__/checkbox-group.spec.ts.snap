// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KCheckboxGroup > When it is disabled, the value changes and the status of the checkbox not changes. 1`] = `"<div class="h-full" style="width: 600px;"><button id="group_change_disabled_btn"></button> <div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="false" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="false" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="true" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > When it is disabled, the value of the checkbox can be initialized. 1`] = `"<div class="h-full" style="width: 600px;"><div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="true" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="false" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="true" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > When it is disabled, when a value of the checkbox changes, the group value is not triggered. 1`] = `"<div class="h-full" style="width: 600px;"><div id="group_trigger_value">3</div> <div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="false" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="false" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md k-cur-disabled"><input disabled="" type="checkbox" value="true" hidden=""> <div class="k-checkbox--box k-checkbox--box__disabled animate-ikun-checking"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > When it is not disabled, the value changes and the status of the checkbox changes. 1`] = `"<div class="h-full" style="width: 600px;"><button id="group_change_btn"></button> <div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="true" hidden=""> <div class="k-checkbox--box bg-ikun-main border-ikun-main animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > When it is not disabled, the value of the checkbox can be initialized. 1`] = `"<div class="h-full" style="width: 600px;"><div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="true" hidden=""> <div class="k-checkbox--box bg-ikun-main border-ikun-main animate-ikun-checking"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="true" hidden=""> <div class="k-checkbox--box bg-ikun-main border-ikun-main animate-ikun-checking"><div><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > When it is not disabled, when a value of the checkbox changes, the group value is triggered. 1`] = `"<div class="h-full" style="width: 600px;"><div id="group_trigger_value"></div> <div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both, __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both, __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > When the disabled value changes, the checkbox value can be synchronized. 1`] = `"<div class="h-full" style="width: 600px;"><button id="group_change_value_btn"></button> <button id="group_change_disabled_btn"></button> <div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="true" hidden=""> <div class="k-checkbox--box bg-ikun-main border-ikun-main animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"><div style="animation: __svelte_185617033_0 200ms linear 0ms 1 both;"><span class="k-icon--base k-icon-transition i-carbon-checkmark text-white" role="" aria-hidden="true"></span></div> </div> 3</label></div></div>"`;

exports[`Test: KCheckboxGroup > props: cls 1`] = `"<div class="k-checkbox-group k-checkbox-group--base k-checkbox-group--test"></div>"`;

exports[`Test: KCheckboxGroup > props: size 1`] = `"<div id="group-sm" class="h-full" style="width: 600px;"><div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--sm"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--sm"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--sm"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 3</label></div></div> <div id="group-md" class="h-full" style="width: 600px;"><div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 2</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 3</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 4</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--md"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 5</label></div></div> <div id="group-lg" class="h-full" style="width: 600px;"><div class="k-checkbox-group k-checkbox-group--base"><label class="k-checkbox--base k-checkbox--base__dark k-checkbox--lg"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 1</label> <label class="k-checkbox--base k-checkbox--base__dark k-checkbox--lg"><input type="checkbox" value="false" hidden=""> <div class="k-checkbox--box animate-ikun-checking"> </div> 2</label></div></div>"`;
